doc/rtd/topics/configuration.rst

```diff
@@ -0,0 +1,89 @@
+.. _configuration:
+
+Configuration Sources
+*********************
+
+Internally, cloud-init builds a single configuration that is then referenced
+throughout the life of cloud-init. The configuration is built from multiple
+sources such that if a key is defined in multiple sources, the higher priority
+source overwrites the lower priority source.
+
+Base Configuration
+==================
+
+From lowest priority to highest, configuration sources are:
+
+* **Hardcoded config**: Config_ that lives within the source of cloud-init
+  and cannot be changed.
+* **Configuration directory**: Anything defined in ``/etc/cloud/cloud.cfg`` and
+  ``/etc/cloud/cloud.cfg.d``.
+* **Runtime config**: Anything defined in ``/run/cloud-init/cloud.cfg``.
+* **Kernel command line**: On the kernel command line, anything found between
+  ``cc:`` and ``end_cc`` will be interpreted as cloud-config user data.
+
+These four sources make up the base configuration.
+
+Vendor and User Data
+====================
+Added to the base configuration are:
+
+* **Vendor data**: :ref:`Data<vendordata>` provided by the datasource
```

Comment:
Should we also reference metadata here too?

Reply:
Is it? What metadata do we fetch from clouds that can be overridden by vendor data?

Reply:
I was thinking specifically of `local_hostname` in metadata is the most likely case where metadata gets overridden by a #cloud-config `hostname` value in either vendor-data or user-data. I can't think of more explicit examples off the top of my head but DataSource does provide a get_config_obj which some datasources (Azure) provide which returns ds.cfg and will participate in the merged config via [ConfigMerger._read_cfg](https://github.com/canonical/cloud-init/blob/main/cloudinit/helpers.py#L263-L282). I think generally it's not frequently a case where that ds.cfg data participates, but there is a mechanism that sets priority at datasource metadata config, vendor data then user data. Not worth holding up this PR as-is, but something we can iterate on more in the future if needed.